Add product query support for Stock indicator block

imanish003 opened the pull request:
Adds Product query support for the atomic stock indicator block(`woocommerce/product-stock-indicator`).

On the client side, when the Stock indicator block is used within the product query block, the markup will be rendered on the server side - No javascript related to Stock indicator block will be rendered.

<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

<!-- Reference any related issues or PRs here -->

Fixes #7333 

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

### Testing
1. Checkout this branch and run npm start
2. Be sure that you have latest version of the Gutenberg feature plugin installed and activated
3. Create a post/page and add the Product Query block.
4. Add Stock Indicator block to Post Template of product query block as shown in video below:

https://user-images.githubusercontent.com/16707866/203531286-f4dd88e8-299e-4bd1-b473-324b35c3c646.mov

5. Now save the post & confirm that:
    1. No JS related to the Product Stock Indicator block is loaded (search in your network tab for product-stock-indicator-frontend.js). As you can see in the screenshot below, product-stock-indicator-frontend.js showing before the changes made in this PR. This shouldn't be visible in network tab.
![Screenshot 2022-11-23 at 4 38 34 PM](https://user-images.githubusercontent.com/16707866/203532085-aea2f3d8-2175-423c-9601-7d8b5c1b94d5.png)
    2. UI of block should be identical in Frontend & Editor.
    3. That the settings are reflected correctly (eg: custom style). Try changing custom style from the sidebar.
    4. ADDITIONAL CSS CLASS(ES)(available in advanced toggle in sidebar) should be added to the container div
    5. Make changes to inventory & test the following status changes on the frontend side depending on the inventory of the product:
        - In stock
        - Out of stock
        - Available on backorder
        - 2 left in stock



* [X] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [X] Experimental

imanish003 committed:
Add product query support for Stock indicator block

On the client side, when the Stock indicator block is used within the product query block, the markup will be rendered on the server side - No javascript related to Stock indicator block will be rendered.

imanish003 committed:
Escape all values in output string

Whenever we are rendering data, we should escape it. Escaping output prevents XSS (Cross-site scripting) attacks.

imanish003 committed:
Merge branch 'trunk' into add/7709-stock-indicator-block-pq-support

Aljullu commented on src/BlockTypes/ProductStockIndicator.php:
```diff
@@ -48,7 +50,88 @@ protected function get_block_type_supports() {
+	 * @param [bool]     $is_in_stock Whether the product is in stock.
+	 * @param [bool]     $is_low_stock Whether the product is low in stock.
+	 * @param [int|null] $low_stock_amount The amount of stock that is considered low.
+	 * @param [mixed]    $is_on_backorder Whether the product is on backorder.
```
Isn't `$is_on_backorder` a bool?

imanish003 replied:
Right, Nice catch! Thanks for pointing this out 🙏

Aljullu commented on src/BlockTypes/ProductStockIndicator.php:
```diff
@@ -48,7 +50,88 @@ protected function get_block_type_supports() {
+		$classnames .= $is_on_backorder ? ' wc-block-components-product-stock-indicator--available-on-backorder ' : '';
+
+		$output  = '';
+		$output .= '<div class="wc-block-components-product-stock-indicator ' . $classnames . '"';
```
Personally, I would prefer to use the `esc_attr()` here than in L120 and L121. Escaping the data right before printing it helps making sure it's escaped; otherwise somebody reviewing the code needs to look at several lines up to see if it's escaped or not.

imanish003 replied:
I agree with you. Thanks for the suggestion. I will make the changes accordingly 🙂

Aljullu reviewed (approved):
LGTM, good job, @imanish003! I left a couple of minor comments, but pre-approving.

imanish003 committed:
Change $is_on_backorder type & escape just before printing

For more info:
https://github.com/woocommerce/woocommerce-blocks/pull/7734#discussion_r1035971939
https://github.com/woocommerce/woocommerce-blocks/pull/7734#discussion_r1035975712

imanish003 committed:
Merge branch 'trunk' into add/7709-stock-indicator-block-pq-support
